src/loaders/xml/client_info.rs

```diff
@@ -0,0 +1,282 @@
+        }
+    }
+
+    pub fn from_name(name: &str) -> Self {
+        match name {
+            "korea" => ServiceType::Korea,
+            "america" => ServiceType::America,
+            "japan" => ServiceType::Japan,
+            "china" => ServiceType::China,
+            "taiwan" => ServiceType::Taiwan,
+            "thai" => ServiceType::Thai,
+            "indonesia" => ServiceType::Indonesia,
+            "philippine" => ServiceType::Philippine,
+            "malaysia" => ServiceType::Malaysia,
+            "singapore" => ServiceType::Singapore,
+            "germany" => ServiceType::Germany,
+            "india" => ServiceType::India,
+            "brazil" => ServiceType::Brazil,
+            "australia" => ServiceType::Australia,
+            "russia" => ServiceType::Russia,
+            "vietnam" => ServiceType::Vietnam,
+            "chile" => ServiceType::Chile,
+            "france" => ServiceType::France,
+            "uae" => ServiceType::Uae,
+            _ => ServiceType::Unknown,
+        }
+    }
+}
+
+fn language_type_from_index<'de, D>(deserializer: D) -> Result<Option<ServiceType>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let index: i8 = Deserialize::deserialize(deserializer)?;
+    let service_type = ServiceType::from_index(index);
+
+    match service_type {
+        ServiceType::Unknown => Err(D::Error::custom("invalid service type {index}")),
+        value => Ok(Some(value)),
+    }
+}
+
+fn service_type_from_name<'de, D>(deserializer: D) -> Result<ServiceType, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let name: String = Deserialize::deserialize(deserializer)?;
+    let service_type = ServiceType::from_name(&name);
+
+    match service_type {
+        ServiceType::Unknown => Err(D::Error::custom("invalid service type {name}")),
+        value => Ok(value),
+    }
+}
+
+/// The ClientInfo's ServerType enumerator.
+#[derive(Debug, Clone, Default, Deserialize)]
+pub enum ServerType {
+    #[default]
+    Primary,
+    Sakray,
+    Local,
+    Pk = 5,
+    Unknown = 999,
+}
+
+impl ServerType {
+    pub fn from_name(name: &str) -> Self {
+        match name {
+            "primary" => ServerType::Primary,
+            "sakray" => ServerType::Sakray,
+            "local" => ServerType::Local,
+            "pk" => ServerType::Pk,
+            _ => ServerType::Unknown,
+        }
+    }
+}
+
+fn server_type_from_name<'de, D>(deserializer: D) -> Result<ServerType, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let name: String = Deserialize::deserialize(deserializer)?;
+    let server_type = ServerType::from_name(&name);
+
+    match server_type {
+        ServerType::Unknown => Err(D::Error::custom("invalid server type {name}")),
+        value => Ok(value),
+    }
+}
+
+fn bool_deserializer<'de, D>(d: D) -> Result<bool, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(d)?;
+    match &s[..] {
+        "" => Ok(true),
+        _ => panic!("don't do that"),
```

Comment:
Something like `"boolean tags may not have any data inside"` seems more fitting xD